Krithiga/basic config (#4)
darchr/gem5-skylake-config: results/docs/Memory_benchmarks.md

# Memory Benchmarks

## Benchmarks Used

> The suite of memory benchmarks which is a part of the [university of wisconsin micro-benchmarks suite](https://github.com/VerticalResearchGroup/microbench) and targets specially the L1 cache miss performance can be used to analyse the performance difference between gem5 and real hardware system.

Reviewer: ```suggestion
The suite of memory benchmarks which is a part of the [University of Wisconsin micro-benchmarks suite](https://github.com/VerticalResearchGroup/microbench) and targets specially the L1 cache miss performance can be used to analyze the performance difference between gem5 and a real hardware system.
```

- MC:  Benchmark with load instructions, generates many Conflict Misses.
- MCS: Benchmark with store instructions, generates many Conflict Misses.
- MD : Benchmark containing cache resident linked list transversal.
- MI : Independent memory access stored in static array, this benchmark has good spatial and temporal locality.
- MIM :Independent memory access stored in static array, this benchmark has very bad spatial locality.
- MIM2 :Same as MIM, but it has 2 extra coalescing ops in between.

## gem5 results:

<img src="../images/memory_benchmarks_gem5.png" width="300" height="300">

> The gem5 results look analytically correct, the MI benchmark which is supposed to have the least missrate theoretically does so as seen in the graph. The MIM and MIM2 benchmarks have bad spatial and temporal locality hence have a very high missrate. The MCS and MC benchmark refelct the cache model beahviour when a conflict miss occurs. The MCS benchmark has more missrate than MC because of the differenc ein store bandwidth, store queue entry size and various other factors in our skylake architecture configuration in gem5.

Reviewer 2: I don't think it's analytical correct, these benchmarks should theoretically give low miss rate.

Author: Theoretically, they do have to give high miss-rate for LRU replacement policy, this is what is my understanding from the benchmark.

> The gem5 results look analytically correct, the MI benchmark which is supposed to have the least missrate theoretically does so as seen in the graph. The MIM and MIM2 benchmarks have bad spatial and temporal locality hence have a very high missrate. The MCS and MC benchmark refelct the cache model beahviour when a conflict miss occurs. The MCS benchmark has more missrate than MC because of the differenc ein store bandwidth, store queue entry size and various other factors in our skylake architecture configuration in gem5.

Reviewer: ```suggestion
The gem5 results look analytically correct. The MI benchmark theoretically has the smallest miss rate, which is shown in the graph. The MIM and MIM2 benchmarks have poor spatial and temporal locality. Hence, these microbenchmarks have a high miss rate. The MCS and MC benchmark reflect the cache model beahvior when a conflict miss occurs. The MCS benchmark has a higher miss rate than MC because of the difference in store bandwidth, store queue entry size, and various other factors in our Skylake architecture configuration in gem5.
```

## gem5 vs Hardware values

> As we have mirrored the memory model configuration in gem5 as close as possible to the intel skylake architecture, we expect to see almost similar missrate in gem5 and intel pcm values. 

Reviewer: ```suggestion
As we have mirrored the memory model configuration in gem5 as close as possible to the Intel Skylake architecture, we expect to see a similar miss rate in gem5 and Intel PCM values. 
```

<img src="../images/memory_benchmarks_hw.png" width="300" height="300">

> From the above graph, it is evident that there is some feature in the hardware which makes it oerform well even for benchmarks that tend to have a higher missrate. This is mostly because of the fact that the strideprefecther and prefecthing techniques supported and modeled in gem5 classic cache model is not as good as the ones found in hardware.

Reviewer: ```suggestion
From the above graph, it is evident that there is some feature in the hardware which makes it perform well even for benchmarks that theoretically have a high miss rate.
We believe the stride prefecther and other prefecthing techniques supported and modeled in gem5 classic cache model are not as good as the ones found in hardware.
```

## Other missing features

> There are other features in gem5 cache model that are not immediately available to mode the memory model to represent a real system like intel skylake architecture. 

Reviewer: ```suggestion
There are other features in gem5 cache model that are not immediately available to model the memory system of the  Intel Skylake architecture. 
```
